In [83]:
import tensorflow as tf
tf.test.gpu_device_name()

'/device:GPU:0'

In [1]:
!pip install vaderSentiment

import pandas as pd
import regex as re 
import nltk

from nltk.sentiment.vader import SentimentIntensityAnalyzer
from nltk.corpus import stopwords

nltk.download('vader_lexicon')
nltk.download('stopwords')

     |████████████████████████████████| 133kB 7.4MB/s 
[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


/usr/local/lib/python3.7/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


True

# Preprocessing Lyrics

In [14]:
df = pd.read_csv('1986-2020_songs_2.csv', encoding='utf-8')
for song in range(len(df)):
  orig = df['LYRICS'][song]

  
  orig = orig.replace('\u2005', ' ')
  orig = orig.replace('\u205f', ' ')
  orig = orig.replace('\u200a', ' ')


  # remove capitlization
  orig = orig.lower()

  # remove punctuation
  orig = re.sub(r'\p{P}+', '', orig)

  # remove stopwords
  orig = orig.split(' ')
  filtered = [word for word in orig if word not in stopwords.words('english') and len(word) > 1 and word not in ['na', 'la']]
  filtered = " ".join(filtered)

  # remove identifiers like verse, chorus, etc.
  df['LYRICS'][song] = re.sub(r'[\(\[].*?[\)\]]', '', filtered)


df

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


RANK  ...    GENRE
0        1  ...      Pop
1        2  ...      R&B
2        3  ...      Pop
3        4  ...      R&B
4        5  ...     Rock
...    ...  ...      ...
3236    96  ...  Country
3237    97  ...  Country
3238    98  ...      Rap
3239    99  ...  Hip-Hop
3240   100  ...  Hip-Hop

[3241 rows x 6 columns]

In [15]:
song = df['LYRICS'][2600]
song.count('fuck ')

8

# Sentiment Scores for Lyrics

In [16]:
analyzer = SentimentIntensityAnalyzer()
df['COMPOUND'], df['NEG'], df['NEU'], df['POS'] = 0.0, 0.0, 0.0, 0.0
for song in range (len(df)):
  scores = analyzer.polarity_scores(df['LYRICS'][song])
  df['COMPOUND'][song] = scores.get('compound')
  df['NEG'][song] = scores.get('neg')
  df['NEU'][song] = scores.get('neu')
  df['POS'][song] = scores.get('pos')
df

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A va

RANK                        TITLE  ...    NEU    POS
0        1  That's What Friends Are For  ...  0.474  0.413
1        2              Say You, Say Me  ...  0.707  0.188
2        3                   I Miss You  ...  0.491  0.401
3        4                    On My Own  ...  0.599  0.233
4        5                 Broken Wings  ...  0.537  0.345
...    ...                          ...  ...    ...    ...
3236    96        More Than My Hometown  ...  0.549  0.374
3237    97                Lovin' on You  ...  0.659  0.238
3238    98                     Said Sum  ...  0.709  0.085
3239    99                        Slide  ...  0.811  0.112
3240   100                 Walk Em Down  ...  0.729  0.113

[3241 rows x 10 columns]

# Profanity Analysis

In [17]:
hip_rap = ['Hip-Hop', 'Rap']
hiphop_rap_songs = df.loc[df['GENRE'].isin(hip_rap)]
len(hiphop_rap_songs)

754

754 out of 3241 songs are labeled as either 'Hip-Hop' or 'Rap'

In [34]:
# profanity vocabulary
profanity = ['nigga ', 'shit ', 'fuck ', 'bitch ', 'niggas ', 'ass ', 'fucking ', 'fucks ', 'niggaz ', 'bitches ', 'sex ', 'dick ', 'pussy ', 'fucked ', 'motherfucker ', 'fuckin ', 'lust ', 'bum ', 'motherfuckers ', 'cock ', 'shite ', 'butt ', 'whore ', 'hoe ', 'hoes ', 'slut ', 'motherfucking ', 'piss ', 'cum ']
df[profanity] = 0 
# collect counts of each profanity word for each song
for song in range(len(df)):
  lyric = df['LYRICS'][song]
  for cuss in profanity:
    df[cuss][song] = lyric.count(cuss) 

df = df.loc[:, (df != 0).any(axis=0)] # remove columns for profanity words that don't appear in any of the songs

df['profanity total']= df.iloc[:, 10:37].sum(axis=1) # sum inidividual profanity columns into new column
df.head(15)

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:659: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[k] = np.nan
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#r

RANK                        TITLE  ... cum   profanity total
0      1  That's What Friends Are For  ...    0                0
1      2              Say You, Say Me  ...    0                0
2      3                   I Miss You  ...    0                0
3      4                    On My Own  ...    0                0
4      5                 Broken Wings  ...    0                0
5      6              How Will I Know  ...    0                0
6      7           Party All the Time  ...    0                0
7      8                Burning Heart  ...    0                0
8      9                        Kyrie  ...    0                0
9     10             Addicted to Love  ...    0                0
10    11     The Greatest Love of All  ...    0               25
11    12                Secret Lovers  ...    0                0
12    13           Friends and Lovers  ...    0                0
13    14                Glory of Love  ...    0                0
14    15               West End Girls  ...    0                1

[15 rows x 38 columns]

In [35]:
df.to_csv('song_data.csv', encoding='utf-8')